### PR TITLE
MKLDNN Backward op cache

### PR DESCRIPTION
**Pull request by ZhennanQin**
Hi all,

For MKLDNN, creating operator primitive and corresponding memory will spend a lot of time, for many scenarios,  the created primitive can be reused if it meets requirement for next computing. In this PR, we implemented the caching mechanism for most backward operators, just like the optimization have done for forward operator.  Correctness test and accuracy test are all PASS. Performance test shown that most models can get speed up in training. Please review this, Thanks. 

@zheng-da, @marcoabreu, @azai91, @TaoLv, @pengzhao-intel

This PR covers below commits.
Enable primitive allocation cache for _backward_LRN
Enable primitive allocation cache for _backward_Pooling.
Enable primitive allocation cache for _backward_Activation.
Enable primitive allocation cache for _backward_Deconvolution.
Enable primitive allocation cache for _backward_BatchNorm.
Enable primitive allocation cache for _backward_Fully_Connected.
Enable primitive allocation cache for _backward_Convolution.

**Correntness test result:**

unit_test - PASS

Train model with dummy data when MXNET_MKLDNN_DEBUG=1

model | Status
-- | --
alexnet | PASS
googlenet | PASS
vgg-16 | PASS
vgg-19 | PASS
inception-bn | PASS
inception-v3 | PASS
resnet-50 | PASS
resnet-152 | PASS

**Accuracy test result:**

CiFAR10 + ResNet50 ( Convergence ): Validation accuracy of 99 epochs ( top1: 67.88%, top5: 96.26%)

CiFAR10 + VGG16 ( Convergence ): Validation accuracy of   74 epochs ( top1: 82.56%, top5: 98.52% )

**Performance test result:**

**Skylake 8180 2sockets training BS=32 (img/s)**

Model | Baseline | After Caching Backward Op | Diff%
-- | -- | -- | --
alexnet | 366.0699 | 412.3794 | 112.65%
googlenet | 104.3128 | 152.8598 | 146.54%
vgg-16 | 38.40072 | 39.65964 | 103.28%
vgg-19 | 32.35834 | 33.80966 | 104.49%
inception-bn | 85.71482 | 124.4676 | 145.21%
inception-v3 | 43.0482 | 57.82697 | 134.33%
resnet-50 | 53.15957 | 65.68513 | 123.56%
resnet-152 | 21.67982 | 28.26535 | 130.38%
inception-v4 | 23.82474 | 29.33314 | 123.12%

**Total Time consumption breakdown(ms)**

Model | Baseline | After caching  Backward Op | Diff%
-- | -- | -- | --
Convolution | 19440.11 | 19441.49 | 100.01%
Activation | 11253.48 | 9493.82 | 84.36%
LRN | 43.325 | 44.281 | 102.21%
Pooling | 1002.97 | 1037.459 | 103.44%
Flatten | 36.84 | 37.138 | 100.81%
FullyConnected | 1135.342 | 1063.388 | 93.66%
Dropout | 210.459 | 207.743 | 98.71%
SoftmaxOutput | 6.944 | 6.428 | 92.57%
_backward_SoftmaxOutput | 3.627 | 4.424 | 121.97%
_backward_FullyConnected | 1792.938 | 1806.433 | 100.75%
_backward_Dropout | 3.977 | 3.586 | 90.17%
_backward_Activation | 11083.87 | 9536.887 | 86.04%
_zeros | 184.517 | 175.302 | 95.01%
_backward_copy | 1.076 | 0.917 | 85.22%
_backward_Pooling | 2291.866 | 2056.797 | 89.74%
_backward_Convolution | 63134.26 | 46223.86 | 73.22%
_backward_LRN | 185.551 | 130.467 | 70.31%
Concat | 594.403 | 618.61 | 104.07%
_backward_Concat | 703.62 | 708.137 | 100.64%
add_n | 2501.635 | 2556.075 | 102.18%
BatchNorm | 5815.348 | 5797.396 | 99.69%
_backward_BatchNorm | 10238.95 | 7818.107 | 76.36%
_copy | 339.458 | 351.205 | 103.46%
elemwise_add | 909.67 | 925.058 | 101.69%

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
ping @zheng-da

**Comment by zheng-da** [comment]
@ashokei could you also help review this PR?

File: src/operator/nn/lrn.cc [hunk 1]
```diff
@@ -129,6 +129,8 @@ void LRNComputeExCPU(const nnvm::NodeAttrs &attrs,
     MKLDNN_OPCHECK_INIT(false, 1, inputs, outputs);
     MKLDNNLRNForward(ctx, param, inputs[0], req[0], outputs[0]);
     MKLDNN_OPCHECK_RUN(LRNCompute<cpu>, attrs, ctx, inputs, req, outputs);
+    // Copy outputs[1] from opcheck reference as backward check needs it.
+    MKLDNN_OPCHECK_COPY_RESULT(outputs, std::vector<size_t>{1});
```
Comment (zheng-da) [inline src/operator/nn/lrn.cc]: do we have to it for all operators?

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 2]
```diff
@@ -175,6 +175,100 @@ void MKLDNNActivationForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
...
+    DType alpha = 0;
+    mkldnn::eltwise_forward::desc fw_desc(mkldnn::prop_kind::forward_training,
+                                          alg, data_md, alpha);
+    mkldnn::eltwise_forward::primitive_desc fw_pdesc(fw_desc, cpu_engine);
+    mkldnn::eltwise_backward::desc bw_desc(alg, diff_md, data_md, alpha);
+    mkldnn::eltwise_backward::primitive_desc bw_pdesc(bw_desc, cpu_engine,
+                                                      fw_pdesc);
+    return bw_pdesc;
+  });
+  LOG(INFO) << "Unsupported data type for MKLDNN activation";
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: I think you should use LOG(FATAL) here.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 3]
```diff
@@ -175,6 +175,100 @@ void MKLDNNActivationForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
...
+  mkldnn::eltwise_forward::desc fw_desc(mkldnn::prop_kind::forward_training,
+                                        alg, data_md, 0.0);
+  mkldnn::eltwise_forward::primitive_desc fw_pdesc(fw_desc, cpu_engine);
+  mkldnn::eltwise_backward::desc bw_desc(alg, diff_md, data_md, 0.0);
+  mkldnn::eltwise_backward::primitive_desc bw_pdesc(bw_desc, cpu_engine,
+                                                    fw_pdesc);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: the code shouldn't run here, right? can we return an empty descriptor?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 4]
```diff
@@ -283,6 +283,157 @@ void MKLDNNConvolutionForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx
...
+  MKLDNNConvBackward(
+      const ConvolutionParam &param, const NDArray &data,
+      const NDArray &weights, const NDArray *bias, const NDArray &output,
+      const mkldnn::convolution_forward::primitive_desc &fwd_pd):
+      bwdData_pd(GetConvBwdData(param, data, weights, output, fwd_pd)),
+      bwdWeights_pd(GetConvBwdWeights(param, data, weights, bias, output, fwd_pd)) {
+  }
+
+  void SetDataNewMem(const mkldnn::memory &out_grad, const mkldnn::memory &weight,
+                 const mkldnn::memory &in_grad) {
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: indent

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 5]
```diff
@@ -283,6 +283,157 @@ void MKLDNNConvolutionForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx
...
+    if (this->bwd_data == nullptr)
+      this->bwd_data = std::shared_ptr<mkldnn::convolution_backward_data>(
+        new mkldnn::convolution_backward_data(
+          this->bwdData_pd, mkldnn::primitive::at(*this->out_grad),
+          mkldnn::primitive::at(*this->weight), *this->in_grad));
+  }
+
+void SetWeightNewMem(const mkldnn::memory &data,
+                       const mkldnn::memory &out_grad,
+                       const mkldnn::memory &in_grad_weight) {
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: indent

File: src/operator/nn/mkldnn/mkldnn_lrn-inl.h [hunk 6]
```diff
@@ -133,13 +134,23 @@ void MKLDNNLRNFwd::_Init(const LRNParam &param,
 
 void MKLDNNLRNFwd::SetDataHandle(const NDArray &in_data,
                                  const NDArray &out_data) {
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_lrn-inl.h]: why `in_data` is not used?

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc [hunk 7]
```diff
@@ -82,6 +82,101 @@ inline static mkldnn::inner_product_backward_weights::primitive_desc GetIPBwdWei
   }
 }
 
+class MKLDNNFullyConnectForward {
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: could you take the caching of FullyConnected forward out and submit another PR for this?
my concern is that this PR is for caching backward. mixing it with all other backward operators might cause confusing later. thanks

**Comment by zheng-da** [comment]
@ZhennanQin sorry for the late reply. I don't see big problems in the PR.

File: src/operator/nn/lrn.cc (hunk 1 above)
Comment (ZhennanQin) [reply src/operator/nn/lrn.cc]: LRN is a special operator that MKLDNN only generates output[0], but for default cpu backward computing, it requires output[1] as well, making opcheck fails eventually. After copying output[1], this problem can be fixed. 
Maybe you're thinking if it's necessary for all operators. Firstly, we only found this issue on LRN. Secondly, it would make results different before and after enabling opcheck from accuracy losing accumulation. Thirdly, it would hurt opcheck performance a lot. So I prefer not applying this to all operators.

File: src/operator/nn/mkldnn/mkldnn_act.cc (hunk 2 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: Agreed.

File: src/operator/nn/mkldnn/mkldnn_lrn-inl.h (hunk 6 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_lrn-inl.h]: It's used in line 140:
this->SetDataHandle(in_data, out_data_mem);

Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_lrn-inl.h]: It's used in line 140:
this->SetDataHandle(in_data, out_data_mem);

File: src/operator/nn/mkldnn/mkldnn_act.cc (hunk 3 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: Yes, it should be a unreachable code. As it's hard to create a empty descriptor(mkldnn doesn't provide such a function to do that), I think fatal error is enough.

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc (hunk 7 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: OK. I will move FullyConnected forward out this PR.

File: src/operator/nn/lrn.cc (hunk 1 above)
Comment (zheng-da) [reply src/operator/nn/lrn.cc]: thanks for explaining. i just didn't understand why we needed this.

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc (hunk 7 above)
Comment (huangzhiyuan) [reply src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: FullyConnected forward PR see #11611.

**Comment by marcoabreu** [comment]
Sorry, I might be missing something, but I can't find any tests. @zheng-da @azai91 could you check please?

**Comment by ashokei** [comment]
changes look ok for me, great performance improvement! Since these are primarily for perf.optimization i assume existing unit-tests should be ok for functionality checking.

**Comment by ZhennanQin** [comment]
@marcoabreu Tests are covered by 2 parts:
1. The existing unit-tests.
2. Train model after enabling OPCHECK by environment variable MXNET_MKLDNN_DEBUG=1. This will trigger computing for each op twice, one in mkldnn pass, the other one in default cpu pass, then compare the results. If the results are different, it would generate error message.

This PR is like a refactoring work, which doesn't add any new feature or functionality, so above tests are enough.

**Comment by marcoabreu** [comment]
Excellent, thanks a lot!

**Comment by ZhennanQin** [comment]
@zheng-da @marcoabreu Code is refactored based on comments:
1. Remove FC forward change into another PR.
2. Change back to use DispatchMode::kFCompute for FC backward.(We still implemented FC backward op cache in this PR, but won't be used)

If there's no more comment, could you please approve and merge this ASAP? Thanks.

**Comment by ZhennanQin** [comment]
Ping @zheng-da @anirudh2290

**Comment by lupesko** [comment]
Thanks for the contribution @ZhennanQin !
Bouncing again for review: @anirudh2290  @zheng-da  @azai91 @mseth10

**Comment by zheng-da** [comment]
This is reviewed. We agreed to merge this to 1.4 instead of in 1.3.

**Comment by pengzhao-intel** [comment]
@zheng-da The 1.3 branch is cut. I think it's the time to merge the PR into master now.
What's your opinion?

**Comment by zheng-da** [comment]
@pengzhao-intel yes. After you rebase successfully, i'll browse through the code again.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 8]
```diff
@@ -193,37 +287,20 @@ void MKLDNNActivationBackward(const nnvm::NodeAttrs& attrs, const OpContext &ctx
...
-    mkldnn::eltwise_backward::primitive_desc bw_pdesc(bw_desc, cpu_engine,
-                                                      fw_pdesc);
-
-    diff_src_memory = CreateMKLDNNMem(in_grad,
-                                      bw_pdesc.diff_src_primitive_desc(), req);
-    stream->RegisterPrim(mkldnn::eltwise_backward(bw_pdesc, *input_mem,
-                                                  *diff_dst_memory,
-                                                  *diff_src_memory.second));
-  });
+  TmpMemMgr::Get()->Init(ctx.requested[activation::kTempSpace]);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: why do you move it here?

**Comment by zheng-da** [comment]
everything else looks fine to me.

File: src/operator/nn/mkldnn/mkldnn_act.cc (hunk 8 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_act.cc]: Can't remember the details, maybe it's changed by mistake. I've changed it back.

**Comment by lupesko** [comment]
@azai91 @eric-haibin-lin @anirudh2290 please review. This has been out for review for a long time...
